### ISSUE TITLE
[Clang-Tidy] `misc-const-correctness` False positive on a variable that cannot be `const`

### ISSUE THREAD

**Issue by TedLyngmo** [llvm/llvm-project#74842]
```c++
#include <fstream>

template<class Archive>
void save(Archive& archive) {
    std::ofstream output("file");
    output << archive;
}

int main() {
    double const dval = 3.14159;
    save(dval);
}
```
`clang-tidy` suggests that `output` could be declared `const`:
```
const.cpp:5:5: error: variable 'output' of type 'std::ofstream' (aka 'basic_ofstream<char>') can be declared 'const' [misc-const-correctness,-warnings-as-errors]
    5 |     std::ofstream output("file");
      |     ^
      |                   const
```
